### PR TITLE
Fix object slice

### PR DESCRIPTION
**Pull request by cyyever**
To avoid casting Tensor to Tensorbase

### CODE REVIEW

File: functorch/csrc/dim/dim.cpp
Comment (Skylion007) [inline functorch/csrc/dim/dim.cpp:870]: Ugh, we do not have the clang-tidy rule enabled for this? Or is it just that it's not flagging except for newer Clang-Tidy?

Comment (cyyever) [reply functorch/csrc/dim/dim.cpp:870]: Detected from LSAN

Comment (Skylion007) [reply functorch/csrc/dim/dim.cpp:870]: I meant for detecting the slicing. Why is this flagged in our current clang-tidy config via [cppcoreguidelines-slicing](https://clang.llvm.org/extra/clang-tidy/checks/cppcoreguidelines/slicing.html#cppcoreguidelines-slicing). Is it because clang-tidy is to old?

Comment (cyyever) [reply functorch/csrc/dim/dim.cpp:870]: From lsan

Comment (cyyever) [reply functorch/csrc/dim/dim.cpp:870]: @Skylion007 Sometimes due to merge race conditions, clang-tidy may not have chance to check them.

File: torch/csrc/autograd/python_variable.cpp
Comment (Skylion007) [inline torch/csrc/autograd/python_variable.cpp:2012]: Why is this changed to a TensorBase from a Variable?

File: functorch/csrc/dim/dim.cpp
Comment (Skylion007) [reply functorch/csrc/dim/dim.cpp:870]: Clang-tidy runs on all files though, not just changed ones though so that shouldn't affect it?

File: torch/csrc/autograd/python_variable.cpp
Comment (cyyever) [reply torch/csrc/autograd/python_variable.cpp:2012]: A chain of interface changes made me found that it actually only needs  const at::TensorBase&

**Review by Skylion007 — APPROVED** [review]

**Comment by huydhn** [comment]
@cyyever Just a head-up that this change requires a corresponding update on https://github.com/facebookresearch/fairseq2/pull/843.  As the PR is ready there, let's see if I can find the code owner to review it to unblock your change here.  

I might need to revert this one though if the change to fairseq2 is not approved in time as keeping github and fbcode out of sync for long will lead to conflicts.

**Pull request closed without merging**
